Source/Public/New-BicepMarkdownDocumentation.ps1

```diff
@@ -0,0 +1,170 @@
+        [Parameter(ParameterSetName = 'FromFolder')]
+        [switch]$Console,
+
+        [Parameter(ParameterSetName = 'FromFile')]
+        [Parameter(ParameterSetName = 'FromFolder')]
+        [switch]$Force
+    )
+
+    function NewMDTableHeader {
+        param(
+            [string[]]$Headers
+        )
+
+        $r = '|'
+        foreach ($Head in $Headers) {
+            $r += " $Head |"
+        }
+        
+        $r = "$r`n|"
+        
+        1..($Headers.Count) | foreach {
+            $r += "----|"
+        }
+
+        $r = "$r`n"
+        
+        $r
+    }
+
+    switch ($PSCmdLet.ParameterSetName) {
+        'FromFile' { 
+            $FileCollection = @((Get-Item $File)) 
+        }
+        'FromFolder' { 
+            $FileCollection = Get-ChildItem $Path *.bicep -Recurse:$Recurse
+        }
+    }
+
+    Write-Verbose "Files to process:`n$($FileCollection.Name)"
+
+    $MDHeader = @'
+# {{SourceFile}}
+
+[[_TOC_]]
+
+'@
+
+    foreach ($SourceFile in $FileCollection) {
+        $FileDocumentationResult = $MDHeader.Replace('{{SourceFile}}', $SourceFile.Name)
+        
+        $MDProviders = NewMDTableHeader -Headers 'Type', 'Version'
+        $MDResources = NewMDTableHeader -Headers 'Name', 'Link', 'Location'
+        $MDParameters = NewMDTableHeader -Headers 'Name', 'Type'
+        $MDVariables = NewMDTableHeader -Headers 'Name', 'Value'
+        $MDOutputs = NewMDTableHeader -Headers 'Name', 'Type', 'Value'
+
+        $BuildObject = (Build-BicepNetFile -Path $SourceFile.FullName).Template | ConvertFrom-Json
+
+#region Add providers to MD output
+        foreach ($provider in $BuildObject.resources) {
+            $MDProviders += "| $($Provider.Type) | $($Provider.apiVersion) |`n"
+        }
+        $FileDocumentationResult += @"
+
+## Providers
+
+$MDProviders
+"@
+#endregion
+
+#region Add Resources to MD output
+        foreach ($Resource in $BuildObject.resources) {
+            try {
+                $URI = Get-BicepApiReference -Type "$($Resource.Type)@$($Resource.apiVersion)" -ReturnUri -Force
+            }
+            catch {
+                # If no uri is found this is the base path for template
+                $URI = 'https://docs.microsoft.com/en-us/azure/templates'
+            }
+            $MDResources += "| $($Resource.name) | [$($Resource.Type)@$($Resource.apiVersion)]($URI) | $($Resource.location) |`n"
+        }
+$FileDocumentationResult += @"
+
+## Resources
+
+$MDResources
+"@
+#endregion
+
+#region Add Variables to MD output
+        if ($null -eq $BuildObject.variables) {
+            $MDVariables = 'n/a'
+        }
+        else {
+            $VariableNames = ($BuildObject.variables | Get-Member -MemberType NoteProperty).Name
+            foreach ($var in $VariableNames) {
+                $Param = $BuildObject.variables.$var
+                $MDVariables += "| $var | $Param |`n"
+            }
+        }
+$FileDocumentationResult += @"
+
+## Variables
+
+$MDVariables
```

Comment:
Can we move the headers around to have it in the order: parameters, variables, outputs?

Reply:
done and done. We may add this as sortable input in some way as well, but lets see if its necessary first.